streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -239,7 +239,7 @@ private void restorePartition(final ConsumerRecords<byte[], byte[]> allRecords,
         final long pos = processNext(allRecords.records(topicPartition), restorer, endOffset);
         restorer.setRestoredOffset(pos);
         if (restorer.hasCompleted(pos, endOffset)) {
-            if (pos > endOffset + 1) {
+            if (pos > endOffset) {
```

Comment:
At this point, we hit a race condition for the restore case of a regular changelog topic: `restorer.hasCompleted(pos, endOffset)` might return `true` even if some record got appended to the changelog iff (by chance), the last `poll()` batch of records ends at "endOffsets - 1".

We should add an additional check here, and get the `endOffset` of the partition again, and check if `restorer.hasCompleted(pos, newEndOffset) == true` -- if it's false, we should also throw `TaskMigratedException`.

Reply:
Maybe I don't recall all the details of the offline conversation, or I'm missing something else, but by fetching the `endOffset` again and doing another `restorer.hasCompleted(pos, newEndOffset)` call won't this cause issues for `KTable` with a source topic?

Reply:
I think @mjsax 's comment was that we can add another `endOffset` call and check if the returned value is still the same, this will help reduce the likelihood of race condition but not eliminate it: however, we need to distinguish the case for source KTable such that in the case when `offset limit ` is specified, we do not need this additional check.

Reply:
ack